File: backend/models/Campaign.js
const mongoose = require("mongoose");

const campaignSchema = new mongoose.Schema(
  {
    ngoId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "NGO",
      required: true,
    },
    title: {
      type: String,
      required: true,
      trim: true,
    },
    description: {
      type: String,
      required: true,
    },
    shortDesc: {
      type: String,
      default: "",
    },
    category: {
      type: String,
      required: true,
      default: "General",
    },
    requiredSkills: {
      type: [String],
      default: [],
    },
    location: {
      type: String,
      required: true,
      trim: true,
    },
    coordinates: {
      lat: { type: Number },
      lng: { type: Number },
    },
    date: {
      type: String,
      required: true,
    },
    startTime: {
      type: String,
      default: "08:00 AM",
    },
    endTime: {
      type: String,
      default: "04:00 PM",
    },
    banner: {
      type: String,
      default: "https://images.unsplash.com/photo-1559027615-cd4628902d4a?w=800",
    },
    capacity: {
      type: Number,
      default: 50,
    },
    volunteersJoined: {
      type: Number,
      default: 0,
    },
    status: {
      type: String,
      enum: ["Active", "Upcoming", "Completed", "Cancelled"],
      default: "Active",
    },
    registeredUsers: [
      {
        type: mongoose.Schema.Types.ObjectId,
        ref: "User",
      },
    ],
  },
  {
    timestamps: true,
  }
);

const { resolveLocationCoordinates } = require("../utils/aiSync");

campaignSchema.pre("save", function (next) {
  if (this.isModified("location") || !this.coordinates || typeof this.coordinates.lat !== "number") {
    this.coordinates = resolveLocationCoordinates(this.location);
  }
  next();
});

module.exports = mongoose.model("Campaign", campaignSchema);

File: ai-service/volunteer-matching/models/campaign.py
from __future__ import annotations

from typing import List, Optional, Union
from pydantic import BaseModel, Field

class Campaign(BaseModel):
    id: str = Field(..., alias="_id")
    title: str
    description: Optional[str] = ""
    category: Optional[str] = "General"
    location: Optional[str] = ""
    coordinates: Optional[dict] = None
    lat: Optional[float] = None
    lng: Optional[float] = None
    required_skills: List[str] = Field(default_factory=list, alias="requiredSkills")
    start_date: Optional[str] = None
    end_date: Optional[str] = None
    date: Optional[str] = None
    maximum_volunteers: int = Field(default=50, alias="capacity")
    current_volunteers: int = Field(default=0, alias="volunteersJoined")
    status: Optional[str] = "Active"
    metadata: Optional[dict] = None

    class Config:
        populate_by_name = True
        extra = "ignore"
